blog/content/post/2024-03-15-uefi-sb-implementation-mpl.md

```diff
@@ -0,0 +1,361 @@
+
+The conclusions allowed us to proceed with UEFI Secure Boot integration in the
+Yocto layer.
+
+## Integrating Secure Boot into a Yocto layer
+
+Integrating UEFI Secure Boot into an existing Yocto layer is possible by using
+the [meta-secure-core](https://github.com/Wind-River/meta-secure-core) layer in
+your build. Its sublayer – `meta-efi-secure-boot` introduces mechanisms that
+allow verifying various files used in the boot process. It offers two secure
+boot technologies:
+
+- UEFI Secure Boot, which verifies images loaded by UEFI firmware against
+certificates
+- MOK (Machine Owner Key) Secure Boot, which extends UEFI SB by
+introducing user-added Machine Owner Keys
+
+### Custom certificates
+
+UEFI Secure Boot allows the user to enroll custom certificates which will be
+used to verify images. The certificates can be loaded automatically thanks to
+the
+[Automatic Certificate Provisioning](https://github.com/Wind-River/meta-secure-core/tree/master/meta-efi-secure-boot#automatic-certificate-provision)
+procedure provided by `meta-secure-core`. The procedure uses `LockDown.efi`,
+which can only be executed when UEFI Secure Boot is disabled. `LockDown.efi`
+provisions the new certificates, which are built into it during build.
+
+UEFI Secure Boot uses the DB certificate to verify GRUB. Aside from that,
+`grubenv` and `grub.cfg` are signed during build using a boot key, which can
+be defined in the layer. GRUB then tries to load the kernel, which is also
+checked against DB. If any of the checks fail, the respective file cannot be
+booted. This mechanism increases the system's security by making sure that the
+firmware does not boot untrusted files.
+
+### Implementation
+
+To integrate the mechanisms available in `meta-secure-core` you need to
+integrate that layer into your build. We use
+[kas-container](https://github.com/siemens/kas/blob/master/kas-container) to set
+up bitbake projects.
+
+In your kas configuration file add the following to the `repos` section:
+
+```yaml
+meta-secure-core:
+url: https://github.com/Wind-River/meta-secure-core.git
+refspec: 8dc9f1b4a735eccee65a8896760e473e110d147e
+layers:
+    meta-efi-secure-boot:
+    meta:
+    meta-signing-key:
+```
+
+Define `meta-secure-core` variables in your layer's `local.conf`:
+
+```conf
+# UEFI Secure Boot variables
+# Use only UEFI Secure Boot without MOK Secure Boot
+UEFI_SELOADER = "0"
+GRUB_SIGN_VERIFY = "1"
+UEFI_SB = "1"
+MOK_SB = "0"
+
+# We want grub-efi from meta-efi-secure-boot to install bootfiles under
+# /boot/EFI/BOOT
+EFI_BOOT_PATH = "/boot/EFI/BOOT"
+
+DISTRO_FEATURES_NATIVE:append = " efi-secure-boot"
+DISTRO_FEATURES:append = " efi-secure-boot modsign"
+MACHINE_FEATURES_NATIVE:append = " efi"
+MACHINE_FEATURES:append = " efi"
+
+DEBUG_FLAGS:forcevariable = ""
+IMAGE_INSTALL:append = " kernel-image-bzimage"
+```
+
+Also, make sure to install the `efi-secure-boot` packagegroup into your image:
+
+```bitbake
```

Comment:
```
the package provides additional features to the
`grub-efi` recipe
```

is this correct? What features does it provide for grub-efi recipe?  From what I see it installs e.g. mokutil which then can be used on booted system to verify UEFI Secure Boot state

Reply:
If I understand correctly, the packagegroup provides a bbappend to `grub-efi`:
https://github.com/Wind-River/meta-secure-core/blob/master/meta-efi-secure-boot/recipes-base/packagegroups/packagegroup-efi-secure-boot.bb#L13

One of the new features is signing `grub.cfg`:
https://github.com/Wind-River/meta-secure-core/blob/master/meta-efi-secure-boot/recipes-bsp/grub/grub-efi-efi-secure-boot.inc#L167

But you're right that I can also mention mokutil.

Reply:
> If I understand correctly, the packagegroup provides a bbappend to grub-efi:

@PLangowski no, packagegroup is (as the name just says), a group of packages which then later can be installed on the image. Here this group consist of packages that are listed here https://github.com/Wind-River/meta-secure-core/blob/master/meta-efi-secure-boot/recipes-base/packagegroups/packagegroup-efi-secure-boot.bb#L13

Reply:
Ok, so it would seem that this packagegroup is not needed, right? We used it in our project, but you don't acrually need any of the packages to integrate UEFI SB.

Reply:
@PLangowski grub-efi is needed, and e.g. mokutil to verify sb state so I would install it

Reply:
https://github.com/3mdeb/news-and-ideas/pull/590/commits/659fe62a534b49d3a97aeb1d9a84e608fcbf93b6